### PR TITLE
Implementation of length penalty using logits processors

### PR DESCRIPTION
**Pull request by maxdebayser**
I realized that I  forgot to sign off all the little commits to fix the formatting.... So I squashed all the commits in this new PR. This includes the fixes proposed in the review comments.

### CODE REVIEW

File: vllm/tgis_utils/logits_processors.py [hunk 1]
```diff
@@ -14,3 +15,19 @@ def __call__(self, token_ids: List[int],
         # transformers warpers assume tensors of shape (batch_size, vocab_size)
         # and the typical warper doesn't use input_ids
         return self.warper(input_ids=None, scores=logits.reshape((1, -1)))
+
+
+class LengthPenaltyWarper:
```
Comment (njhill) [inline vllm/tgis_utils/logits_processors.py:20]: @maxdebayser apologies, I know I suggested making use of the transformers one but after looking at it I don't think that's a good idea (especially having to create a new tensor), and kind of remember why we had our own logic in TGIS!

I think something closer to your original impl in the prior PR would be better, but incorporate the handling of negative logits, so something like:

```python
    def __init__(self, length_penalty: Tuple[int, float], eos_token_id: int):
        self.start, self.penalty = length_penalty
        self.eos_token_id = eos_token_id

    def __call__(self, token_ids: List[int], logits: torch.Tensor) -> torch.Tensor:
        if (tokens_past := len(token_ids) - self.start) > 0:
            eos = self.eos_token_id
            increase = torch.abs(logits[eos]) * (pow(self.penalty, tokens_past) - 1)
            logits[eos] += increase
        return logits
 ```
 
I think the negative handling fix was made since we implemented this in TGIS so it would be great to update that in line with this too!

**Review by njhill — APPROVED** [review]
Thanks @maxdebayser!

**Comment by maxdebayser** [comment]
That's strange, seem like the logits tensor has a different shape in Mixtral. I didn't see anything like that with Mistral-7b, I'll have to take a look. But it happens only for the requests with length_penalty, right?

**Comment by njhill** [comment]
@maxdebayser actually it turned out to be a preexisting bug in the typical p logits processor. But there was also a bug where we weren't checking for presence of the param properly so the length pen logits checker was always getting added. Fixes here: 2961da7ff4213403ba68acddab48f3efb44a2f90.

**Comment by maxdebayser** [comment]
That explain's why I wasn't able to reproduce it :D. I tried with a quantized Mixtral 7Bx2 on my machine and was about to launch the full Mixtral 7Bx8 to test.
